### DISCUSSION TITLE
Referencing issues

### DISCUSSION THREAD

**Discussion by idiosyncrasy00** [eclipse-langium/langium#1015] [Q&A] [answered]
Hi, I am writing the grammar for the DSL language and making a langium-client to generate from my DSL into text. The part of grammar is expressed as below:
```
UsecaseCS:
         'usecase' name=UnrestrictedName
         ('description' '=' description=STRING)?
	'primaryActor' '=' primaryActor=GetActorCSName
  	('secondaryActors' '=' '{' secondaryActors+=GetActorCSName (',' secondaryActors+=GetActorCSName)* '}')?
	(precondition=UsecasePreconditionCS)?
	(postcondition=UsecasePostconditionCS)?
	firstStep=StepCS
         (extPoints+=ExtensionPointCS)*
	'end'
;

StepCS:
	(stepCS1=UCStepCS | stepCS2=RejoinStepCS | stepCS3=ActStepCS)
	(altFlows+=AltFlowCS)*
	(nextStep=StepCSTemp)?
;

StepCSTemp:
	(stepCS1=UCStepCS | stepCS2=RejoinStepCS | stepCS3=ActStepCS)
	(altFlows+=AltFlowCS)*
	(nextStep=StepCS)?
;

ActStepCS:
	ActorStepCS | SystemStepCS
;

ActorStepCS:
	(isActorStep?='actStep') (name=UnrestrictedName)?
	('description' '=' description=STRING)?
	'from'
		preSnapshot=SnapshotPatternCS
	'to'
		postSnapshot=SnapshotPatternCS
	( 'actions' (actions+=ActorActionCS)* )?
	'end'
;

SystemStepCS:
	'sysStep' (name=UnrestrictedName)?
	('description' '=' description=STRING)?
	'from'
		preSnapshot=SnapshotPatternCS
	'to'
		postSnapshot=SnapshotPatternCS
	( 'actions'
		(actions+=SystemActionCS ';')*
	)?
	'end'
;

RejoinStepCS:
	'rejoinStep' rejoinTo=UnrestrictedName
	('description' '=' description=STRING)?
	('when'
		condition=SnapshotPatternCS)?
	'end'
;

UCStepCS:
	'inclStep' (name=UnrestrictedName)?
	('description' '=' description=STRING)?
	includedUC=[UsecaseCS:QualifiedName]
	'end'
;

AltFlowCS:
	'altStep' (name=UnrestrictedName)? 'at' baseStep=UnrestrictedName
		('description' '=' description=STRING)?
		'when'
			condition=SnapshotPatternCS
		'end'
		altStep=StepCS
	'end'
;
```

The relevant function, built up from AST tree in langium-cli:
```
export function generateUsecaseCS(usecaseCS: UsecaseCS, fileNode: CompositeGeneratorNode): void {
    //other code
    let stepCS1 = usecaseCS.firstStep.stepCS1?.includedUC.$refText;
    let stepCS2 = usecaseCS.firstStep.stepCS2?.rejoinTo;
    let stepCS3 = usecaseCS.firstStep.stepCS3?.name;

    console.log("ucStep is ", stepCS1); //result is always undefined 
    console.log("rejoinStep is ", stepCS2); //result is always undefined
    console.log("actStep is ", stepCS3); //result is printed if the relevant coding block in DSL file is added
}
```
The issue is likely lies in here.
```
(stepCS1=UCStepCS | stepCS2=RejoinStepCS | stepCS3=ActStepCS)
```
I've read the langium documentation about the referencing issues [/discussions/829] and I wonder if my code falls in that scenario. I wonder if the issue is fixed now and there is a solution for this. The textual DSL file works fine without errors, but the code generator does not.

**Comment by idiosyncrasy00** [comment]
P/s: The AST for given above piece of code is generated as follow:
```
export interface UsecaseCS extends AstNode {
    readonly $container: FrslModel | PackageCS;
    readonly $type: 'UsecaseCS';
    firstStep: StepCS
}

export interface StepCS extends AstNode {
    readonly $container: AltFlowCS | StepCSTemp | UsecaseCS;
    readonly $type: 'StepCS';
    altFlows: Array<AltFlowCS>
    nextStep?: StepCSTemp
    stepCS1?: UCStepCS
    stepCS2?: RejoinStepCS
    stepCS3?: ActStepCS
}

export interface StepCSTemp extends AstNode {
    readonly $container: StepCS;
    readonly $type: 'StepCSTemp';
    altFlows: Array<AltFlowCS>
    nextStep?: StepCS
    stepCS1?: UCStepCS
    stepCS2?: RejoinStepCS
    stepCS3?: ActStepCS
}

export type ActStepCS = ActorStepCS | SystemStepCS;

export interface ActorStepCS extends AstNode {
    readonly $container: StepCS | StepCSTemp;
    readonly $type: 'ActorStepCS';
    actions: Array<ActorActionCS>
    description?: string
    isActorStep: boolean
    name?: UnrestrictedName
    postSnapshot: SnapshotPatternCS
    preSnapshot: SnapshotPatternCS
}

export interface SystemStepCS extends AstNode {
    readonly $container: StepCS | StepCSTemp;
    readonly $type: 'SystemStepCS';
    actions: Array<SystemActionCS>
    description?: string
    name?: UnrestrictedName
    postSnapshot: SnapshotPatternCS
    preSnapshot: SnapshotPatternCS
}

export interface RejoinStepCS extends AstNode {
    readonly $container: StepCS | StepCSTemp;
    readonly $type: 'RejoinStepCS';
    condition?: SnapshotPatternCS
    description?: string
    rejoinTo: UnrestrictedName
}

export interface UCStepCS extends AstNode {
    readonly $container: StepCS | StepCSTemp;
    readonly $type: 'UCStepCS';
    description?: string
    includedUC: Reference<UsecaseCS>
    name?: UnrestrictedName
}

export interface AltFlowCS extends AstNode {
    readonly $container: StepCS | StepCSTemp;
    readonly $type: 'AltFlowCS';
    altStep: StepCS
    baseStep: UnrestrictedName
    condition: SnapshotPatternCS
    description?: string
    name?: UnrestrictedName
}
```

**Comment by msujew** [comment]
Hey @LockedAwya,

at least with the grammar you have supplied us, I can't reproduce the issue (see [this playground link](https://langium.org/playground?grammar=OYJwhgthYgBAEgUwDbIPYHU0mQEwFD6IB2ALiAJ6wCqAzogMZj0DCAygFz6w%2B9-%2BwA9ILSkAFojgM0uRLXwBIBQDMAliFqk2pRAAcAvNr3t8AbkJHd7LgoAUm42wCM%2B6ux1W2sAD6wHngCZ9ACVEACs0VWJLdh8-D3YAZn0AQQYtBLYASkVbMGRSADF0AHdaAGpUguK0EvYsgCpc4kQADwy9Q0yAFUQIXSyAfjMLHr7dG3tMlzcYr19-diDQiKi5uMW2ZLSOzxy7fKLSiqqj2vqmuxb2yy7HIZH8HbmbHex13zYKBwg5x7eQC9cqpaADLIN9AByMDpSyQrKwWzESCIfQASQAIg87JDZLQGCBVLpSKo0MRIbBIVDYHiCUSSWT0VjhgpIcoQGgIJDFApdCBEGxkbpaGJREzFJDSGhuUpdGhNIKwMLRaRxaySLhueZ8F8fkDWbRvnDEciIKjMdjbLi5HTiaTyZTqbTCXbGRaWWyOVyeXyBUKRWLMRKpTLefKtP6VWrIRqtYQVpFopkbJD%2Bask3oKWnE900Gqrc76faKVSKYXXcQmZbISUJOSedJiLhVAzK%2B6JbHHrNkxKogxkMakSiqyyCzaXa2S07x0W3czFH3kABXWS4Nz6ADadEYzEQ1kxAF0O02409qqVrBLDoPTebmZSwKQKQAjXe3INKMf4ifFx1lmcViOPI1nWoYKI2zatmq6onjy14eHcnjHpqjxiKouCyMQsA6CAEBRPksAYJwQgADoVII5g4XhyLILAmIcEIG4APpgAAtAAXikrEAFoHhuJElExB4NBRhBoRhJDYZI1EEQAsgAMkxLAAPKybJACiABy3QMYIJF6Q0-G0CRbDCYMJENPponiZhUm4fhtFsIpKlqVpOmkXpggbgAeiRxAkSAwmiUAA&content=E4UwVg9glgdgygFxABwARIM4NSGATIA)). Langium seems to parse the `stepCS2=RejoinStepCS` assignment as expected:

![image](https://user-images.githubusercontent.com/4377073/230648365-580a7f84-0d55-43cb-b7e2-95624cb08762.png)

The issue is likely somewhere else. Can you take a look at the output window of your language server in vscode? It might print a warning about grammar ambiguity, which could explain the behavior you're seeing. If anything, this is related to the parser part of Langium, not the generated code, which are just interface definitions that get erased at runtime anyway.

**Reply by idiosyncrasy00** [reply]
Yes, the grammar ambiguity exists:
```
Ambiguous Alternatives Detected: <0, 1> in <OPTION1> inside <UsecasePreconditionCS​> Rule,
<description, =, STRING, IDENTIFIER, :> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
Ambiguous Alternatives Detected: <0, 1> in <OPTION3> inside <ObjVarCS​> Rule,
<IDENTIFIER, ;, IDENTIFIER> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
Ambiguous Alternatives Detected: <0, 1> in <OPTION1> inside <UsecasePostconditionCS​> Rule,
<description, =, STRING, IDENTIFIER, :> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
Ambiguous Alternatives Detected: <0, 1> in <OPTION3> inside <ObjVarCS​> Rule,
<Real, =, STRING> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
```
Update and add relevant parts, fyi:
```
ObjVarCS:
	(isMatched?='$')?name=UnrestrictedName ':'
	(anytype=UnrestrictedName ('('anytype1=UnrestrictedName')')?)?
	(type=[ClassCS:QualifiedName])?
;

UsecasePreconditionCS:
	'ucPrecondition'
		('description' '=' description=STRING)?
		snapshot=SnapshotPatternCS
	'end'
;

UsecasePostconditionCS:
	'ucPostcondition'
		('description' '=' description=STRING)?
		snapshot=SnapshotPatternCS
	'end'
;

SnapshotPatternCS:
	('snapshotID' '=' name=UnrestrictedName)?
	('description' '=' description=STRING)?
	//objects+=ObjVarCSWithClassType | objects+=ObjVarCSWithPrimitiveTypes |
	( (objects+=ObjVarCS) ';')*
	(links+=VarLinkCS ';')*
		(constraints+=ConstraintCS)*
;
```

But the given warnings seem to have nothing to do with the aforementioned issue. 
P/s: I also add more information in UsecaseCS definition in the first post, you can take a look on it.

**Reply by msujew** [reply]
Can you just link a reproducible example somewhere? It's kind of hard to make a whole grammar using the snippets you provided with missing terminals and so on. I can't think of a reason why this is happening, given that the grammar that you posted initially works as expected.

**Reply by idiosyncrasy00** [reply]
Sure thing. The editor works fine but I don't know where are the ambiguity rules from and if it is a referencing issue. This could lead to the adversity of extracting textual information from the content. 

[Link to the example](https://langium.org/playground?grammar=IIYwLg9gTgwgygLgFAEgDkBDc00AIB2GAtgKYC8AqvlCQM5hQCW4JAJgHLEmoAUardEEwAOYRhHx40ZPANpDGo8fjJwAKgCUAkuwDiASgD8qNCXys0SANxIkukmFCRYcTqWQoszsgG0n0eAQqGnomFg4uAF0bJApaEhAMeMCkXDT09LQAV3jE%2BLxCUkpqOgZmMDY3bgyavjkFJQkpGVx6kTEJVU0dA2N0YSYiDCgAT38oZrwBxiHR8bJ7R2wXKtTcFD5ciVZhseXaSdw0AG88LfNd8doAagWHcfgq3D4AGjOE7cv928WH1y59AAqI4AXzQRl4Aw%2B5kYHRUcQSSRIAAUaCBtrDlPAIRthBB6OiYXDKLkkcj8WBCaxMRJsX0AGaMKD0OAVYSqNnwExmCzWWyskjCFIbeiC%2BAARko8E5cFwAB9cKKhXAAExkDQkABWEEY%2BAFyvliplAGYyE59djeBgADZgABi1ogAHcbmbbQ7ndjAbx8CQAB5gfUcsXqEhEYQQmIW0Phjw8JUSqXRw0J1XqrU6vUylMms3gaP6K3ux0u27AYueuBAn3%2BwNs4PKyO2c0yjzjZMKuAjUVEaN8pDt1u8Ri0QeCwwyLz68HPQrkYKlMIVCKkHF1QTtZSHNqKYnqbR6HFoelQCBESwoFBQuCEYS0AAWEDAqlvD6fyIwYAqUHwXPQkAvK8KRvDA70fZ8QLA99P2-X84G5cxLCjbsKl7Id0Fobtp1nLhihCMpwiqNd%2BA3XctyOFod0aFR9x6I8TzPQDr1fcCX1At8wA-L8SB-P80AA1AgJZFinzYqDOJgni4IQ3kYg1bVdWjDw0BoBSs0FPBVMzNQIDwxdykqLheBI%2BRNyaCjZFI6iugPXpjKde8zEAqkaRokSJO43iqz6UxEP7ChpRDZTdRAa1sJ4Oc9NCAyVxIYiqLhbcrL3bpDz6ELrSyARWAC3wETyEhAgARSyG1GEZQzSEiGSkObCsnWFTBbXCyKF2iwiAUMI5PzwAAjJEgzagjl1WS911MsjzOkSyJus2i0sEtAHKcwSUBcvd3K42C-3QHlAJtOtBQbPi9v7AB5XrNQANWGYUeBHABZT8QEc1gJzQAAScFDFakp2pGrgjgQC8eAwfARjAEZhHnP7hsqkhnjQPgwYhqGSElIal3h8FvpxAB6PGeEh6HfBga0kloYrSutcrGHhyIm1iUl4lRaFqThRqshAVn1q3QTxoaRKLNaZLlBsui%2BhQWh3LEjitqkk6-JifKyQpXnaUQEwufJAkMUS-mTMF8jppF2aUtsnEpZlyC5ckryav7G3wPlry4zQaX2PArQABFDl%2B-Csdi%2BLRamyiQ5o1K7I2Z4IEuhIwFdC7rtuqsjiscFvWV2HA6eGgwCyH9aCNJh8AAcw8H2AFF2DULQ7S0SuNCQBUdDUZvMHwfAn0-Pm5XQVgIBALJSHwMAe6aZv0F1AA3YZGDBsBLD7tAae-G0l-QCBhDxWhYRIDf3Z4%2BeaYALwwXrrX3yf3e7IhrVqpASrKirYtwPOC-wIvorLjxMZip4%2BAADo8B-w6quQEMR7yMFYAIfAuBvxEF1DaXAAB1RAuA8YAB0bh4xsAgpB1pcBVxrnXBuGgEAYJ8AAfQwAAWhPsAWhAAtSIPhMFOioZEQEuCkD4MIIQ1ub8HAfyLvgYevUeIULxj4AADLQgAnJEa4PC%2BHIPmroKRAAiHgmDMGALlD4AAepo3RDNASaLlEjXR%2BijFoFMUCNAPCkBQJgWYeBPFEH8NwA9AAMlQmAZ0HoPWrmoKRmCsGAjYbQTBcAuGGEwYCcJPCXGwPcVATxai-EBKCSEsJWDpGGMwfgTBUAuG4KAA&content=IYYwLg9gTgBAwgVwM6QLYFMoCh0DsAmWWy6IwS6MAEsAQDboAKwAnhrmFgJAAOUAlqmBQWAQXDQYAXnjI0mbqDABlMOh4wkARhgAzKBFTdI3PITPEKZCjEYGQ6JEmXAG3fI5ACeYfhFzSMADkAHRhQVxc3HyCwmISsDKIKIYKXEqq6po6%2BobGEKYE3FDoAFYQ-LiZGtowAO4AFniFhFyVIHTV1LT4DMxseJxcFungXUgALHoGRlwmw0XpdCpqGsBg2TDcXI3NkSOjK1nawAD6OTPb8wutFmZAA)

P/s: link was updated. I cannot see the syntax tree in the playground.

**Reply by msujew** [reply]
You're missing an entry rule, but otherwise it seems to work as expected. [Link](https://langium.org/playground?grammar=OYJwhgthYgBAKgUwM4BcBQ7EDtUgJ6wCyA9gCaIA2AXOgJAAUViEOqyA1ALwMCCAxqhIgAwgGVYAH1gBVZIn5h54gJQqAVAG5MAoaLG06AcjCDhR2NkiIuM7CBR4AloMRkActfoMjFZPxAnAAdUJxJsCyMuCz8A4NDwrjF4ACUASXcAcRUAfnojHDIjdG10TMRUXWFxT1ZDUz0uAG0q-Wo7BzRA1w9rAF1SuQUlRHFaWAnJyaMAV3lFeQsrVlt7R27UN1rEdCm9n1jAkLCI2CiYlDjjxOT0rNz8oMDoAlbI6Ngnpxf8Vq5yypmfTbXawRhGebhMgwX5A5DvSIAbwskOw0NecO4ANaNWssB8ABoUQooTDWpx-hUcWJthozgBfIwPRhPElopwJbC2eYjAAKDn4UI5J1UeRZJDQgvZnO5w3kvIlqClZGF4VF9AAZk4QGgxJsgkl9eJ8oViqU9YggmNvGhLeIAIy2cRGiTSW1WsQAJi4KUQACsSE5sBaPVJYO7xABmLi6EOqbxgSioABilBIAHcKbwk6mM6p1N5sIgAB6oEOGu3JFhBB7ml1ICBBQwMCNiR0yZ2VsOt72%2BgNBuOu8Mu6Oxl0qBM5tOZ7jZlPT-OFktl-UVj21nSCQeGVqDsNifC2iCDkqbvTb7xOZC7-U5aINENM-HLGwddYuTa9VjMg6XI6chFYEOeITiSVIMmyMUjA1EASAgYo6DoVkxCsIJkAACxIVAklQjCsN5MBUE2EBsGNYwhAQpDFRQsA0Mw7CaLo-DCOI0ixBNNEzUwA8jwvYxkEPR9n2sVZOmcHpaTFX9-H-E5AOA64uVuCDmWg2D4PoJCHEYvCGNw%2BiCKIxASLIowKM0oJqP0rCcNo3TDNY0zTVPdA%2B0DYMXUMIwHH7DzLQsHz3PgEhRPfCSvHBBSALOD4otA5T7ikox03QnBKOVVUlOs1AHOMtjVOcwZOw9Lyg34SghIYF9Qq6D8tmsH9fD-EDwnk5rFLAu5IPoMrKBmCgyA7ZohgWUYDAARRmRMnC1erWD6Diihcudc3Ta1jETFd-OElY31q8LvxyM5CIsAAjEZyz28TPxBRDpKuaLziA9qZQS7rEOS1KIk0ugMte7LcpM9iPucxDNvLE9jEKzAAHlTr9AA1GB1oYK8iEI-hUrIO8jAAEiZHJqqujY5sQM5qAQhgwGwfBUHwIJXzWfabrxHwfGp2n6cQdsmeu0mmQJ5kAHohYYOmGeaERKCUZAxkm6bZq-RA%2Bg3dARr5AUhU5dbZn4fk2RVADNPu2TWpii4ZJapTwMSzTkGy2ymJyli8qcziXPV%2BVFT%2BkUDHyGY9e9rW5ONprLcUtrw9em33roe27Pox37JdoHFq49AdIMlO2ObCFsrSAARQCid5kmlcauKzaeyvra6n98RIeGFHYbg4cR5GxBUM5NCZAtBlLuqldgBxUBmEjkGHQJsGAQxC4AUXceA0mTNI55SdBpAyeAN5MbBsCwwiQ8kYwyBIfgZlYXBD9ajfjCDAA3GAnGp1BimPoxKA5YzEzf4wSCCSyyAv6-whMZZ%2Bn8ABeYBTqUEQCAgSyAICUHTvLT%2BittjDwqGPbAE9arT0MMTQeGCfAADoLCEIOogDQpR0JODIBQbAsBiIQCDImWAAB1AwsAhYAB1OBC20Mw1hlBYDz0XsvVeKRqDcKaAAfTAAAWggbwBRAAtPoTQeHplkX0dQAj0BCKsCIremDR7j0sBfU6xlpFCyaAABgUQATj6BwfRhi2FvRsQAIgYDwnhJDJBNAAHpeL8SrdQXjJBGF8f4wJQSjBhI0EYfR6BaH0JwEw4yLCjHEAADKyJEDDIgRAF7wBsTw3h6hNHIB4WIXROQeHqAqfotJDDMkgGyR4-JhTimlPKbw2xQSeHYB4SAXRAigA&content=IYYwLg9gTgBAwgVwM6QLYFMoCh0DsAmWWy6IwS6MAEsAQDboAKwAnhrmFgJAAOUAlqmBQWAQXDQYAXnjI0mbqDABlMOh4wkARhgAzKBFTdI3PITPEKZCjEYGQ6JEmXAG3fI5ACeYfhFzSMADkAHRhQVxc3HyCwmISsDKIKIYKXEqq6po6%2BobGEKYE3FDoAFYQ-LiZGtowAO4AFniFhFyVIHTV1LT4DMxseJxcFungXUgALHoGRlwmw0XpdCpqGsBg2TDcXI3NkSOjK1nawAD6OTPb8wutFmZAA)

**Comment by idiosyncrasy00** [comment]
@msujew I just notice a reference complexity in the code thanks to the syntax tree in your playground. The syntax tree does not treat stepCS1, stepCS2 and stepCS3 as the same level, instead each latter declared component is treated as a nested component inside the previous declaration.

The context grammar is: 
```
(stepCS1=UCStepCS | stepCS2=RejoinStepCS | stepCS3=ActStepCS)
(altFlows+=AltFlowCS)*
(nextStep=StepCSTemp)?
```

That is for example in the content:
```
usecase ProcessSale
	inclStep HandlePayment
	end
	actStep s1 from
	to
	end
	rejoinStep s1 when
	end
	actStep s4 from
	to
	end
end
```
The following syntax tree for firstStep is:
```js
{
    firstStep: {
        $type: "StepCS",
        stepCS1: {
            $type: "UCStepCS",
            includedUC: Reference('/elements@1')
        },
        nextStep: {
            $type: "StepCSTemp",
            stepCS3: {
                $type: "ActorStepCS",
                isActorStep: true,
                name: "s1"
            },
            nextStep: {
                $type: "StepCS",
                stepCS2: {
                    $type: "RejoinStepCS",
                    rejoinTo: "s1"
                },
                nextStep: {
                    $type: "StepCSTemp",
                    stepCS3: {
                        $type: "ActorStepCS",
                        isActorStep: true,
                        name: "s4"
                    }
                }
            }
        }
    }
}
```
If I reorder the content a little bit, that is: 
```
usecase ProcessSale
	description = '...'		
	primaryActor = Customer
	actStep s1 from
	to
	end
	rejoinStep s1 when
	end
	inclStep HandlePayment
	end
	actStep s4 from
	to
	end
end
```
The following syntax tree for firstStep is modified as well:
```js
{
    firstStep: {
        $type: "StepCS",
        stepCS3: {
            $type: "ActorStepCS",
            isActorStep: true,
            name: "s1"
        },
        nextStep: {
            $type: "StepCSTemp",
            stepCS2: {
                $type: "RejoinStepCS",
                rejoinTo: "s1"
            },
            nextStep: {
                $type: "StepCS",
                stepCS1: {
                    $type: "UCStepCS",
                    includedUC: Reference('/elements@1')
                },
                nextStep: {
                    $type: "StepCSTemp",
                    stepCS3: {
                        $type: "ActorStepCS",
                        isActorStep: true,
                        name: "s4"
                    },
                }
            }
        }
    }
}
```
I wonder if there is a way to resolve the "nesting problem"?

**Reply by msujew** [reply]
Right, that is kind of nesting in your AST expected, given that there's no indicator for nesting in your language. Usually, DSL/programming languages define some sort of delimiter to identify when a nested element is supposed to start/stop. Usually something like parenthesis, braces or brackets (`()`, `{}` or `[]`) works, although indentation can also work (Python/YAML).

Langium discards all whitespace tokens by default, but it's still possible to build a language that uses whitespace sensitive indentation. For that you will need to override the [`TokenBuilder`](https://github.com/langium/langium/blob/main/packages/langium/src/parser/token-builder.ts) service and implement something like the [Chevrotain Python Tokenizer](https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js) example. You need to use that special `indent/dedent` token in your grammar to signal to the grammar that it should only nest elements if the parser reads the `indent` token, and that it should end the nesting once it reads the `dedent` token.

I'm wondering right now whether we should have some additional functionality in the framework that automatically creates `INDENT`/`DEDENT` tokens based on decorator information. I've created https://github.com/langium/langium/issues/1016 to track this feature.

**Comment by idiosyncrasy00** [comment, marked as answer]
@msujew I figured out to my own references issue. The recursive statement at ```(nextStep=StepCSTemp)?``` which made the AST at ```firstStep``` nested. Simply deleting it and modifying several relating parts solved the problem. The ambiguity have nothing to do with this tho.

Changing from this:
```
(stepCS1=UCStepCS | stepCS2=RejoinStepCS | stepCS3=ActStepCS)
(altFlows+=AltFlowCS)*
(nextStep=StepCSTemp)?

AltFlowCS:
	'altStep' (name=UnrestrictedName)? 'at' baseStep=UnrestrictedName
		('description' '=' description=STRING)?
		'when'
			condition=SnapshotPatternCS
		'end'
		altStep=StepCS
	'end'
;
```

To this:
```
(stepCS1+=UCStepCS | stepCS2+=RejoinStepCS | stepCS3+=ActStepCS)* //(* or + is acceptable)
(altFlows+=AltFlowCS)*

AltFlowCS:
	'altStep' (name=UnrestrictedName)? 'at' baseStep=UnrestrictedName
		('description' '=' description=STRING)?
		'when'
			condition=SnapshotPatternCS
		'end'
		altStep+=StepCS
	'end'
;
```

Note that ```AltFlowCS```  with statement ```altStep+=StepCS``` still gets the recursion and if digging deeper, there are still more repeated errors but it is not a big deal for me. Thank you for your cooperation. Much appreciated.

[Link to the fixed code](https://langium.org/playground?grammar=OYJwhgthYgBAKgUwM4BcBQ7EDtUgJ6wCyA9gCaIA2AXOgJAAUViEOqyA1ALwMCCAxqhIgAwgGVYAH1gBVZIn5h54gJQqAVAG5MAoaLG06AcjCDhR2NkiIuM7CBR4AloMRkActfoMjFZPxAnAAdUJxJsCyMuCz8A4NDwrjF4ACUASXcAcRUAfnojHDIjdG10TMRUXWFxT1ZDUz0uAG0q-Wo7BzRA1w9rAF1SuQUlRHFaWAnJyaMAV3lFeQsrVlt7R27UN1rEdCm9n1jAkLCI2CiYlDjjxOT0rNz8oMDoAlbI6Ngnpxf8Vq5yypmfTbXawRhGebhMgwX5A5DvSIAbwskOw0NecO4ANaNWssB8ABoUQooTDWpx-hUcWJthozgBfIwPRhPElopwJbC2eYjAAKDn4UI5J1UeRZJDQgvZnO5w3kvIlqClZGF4VF9AAZk4QGgxJsgkl9eJ8oViqU9YggmNvGhLeIAIzcGTiI0SaS2q1iABM3BSiAAViQnNgLZ6pLAPeIAMzcXSh1TqbxgSioABilBIAHcKbwU%2Bmswn6AB6IsMbCIAAeqFDhrtyRYQQepRL8frECChhLDEjYntthddfDPa9XD9geDraHrqjXDjrpUxdLybTGezsbzq8LdC75arNdbDxLpTndcMrUn0jE%2BFtEFbJR0QNbhgYTmQ5-1OWiDVDTPxyxsHTrC4my9KwzIHJcRycgisCHPEJxJKkGTZGKRgaiAJAQMUdB0KyYhWEEyAABYkKgSQEcRpG8mAqCbCA2DGsYQjYbhir4WAhEkWR7GcVRNF0QxYgmmiZqYFeN5PvkyDXj%2Bf7WKsnTOD0tJihB-hQScMFwdcXK3MhzJoRhWH0LhDg8ZR3EUVx1G0Yg9GMUYzEmUEbFWaR5EcRZNkCQ5pr3ugY5BiGrqGEYDjjsFloWOFQXwCQClAcpXjgtp0FnB8qUIXp9yqUYmZETgLHKqquluag3l2YJBl%2BYMA6eqFwb8JQsllvJgFdMBWzWOBviQfB4RaX1OmIXcKH0I1lAzBQZDOs0QwLKMBgAIozMmThal1rB9MJRT%2BbmK4FgY%2BTLi1-4JR1SVgTkZw0RYABGIw1u1SkgSCOFqVcaXnLBQ0ytlY04XlBURCZdDFX9ZUVfZQmA35OEnfq3B3sYNWYAA8nd-oAGowNajCvkQNH8AVZCfkYAAkTI5Gdz0bJtiBnNQ2EMGA2D4Kg%2BBBABawXa9eI%2BD4rPs5ziB9rTnWgYgTJU8yXYc1zzQiJQSjIGMK1rRtkt9E2mDzXyApCpyeOzPw-Jsiq0EmR9GkDelFzqf1ulITlJnIGVHm8eV-GVb5In%2BXr8qKuDIpHcYMym0HhuaVbvUOzpg1x39zsA3QbueVxHted70M7aJ6DmdZ2eCc%2BEJlWkAAiME0zzL30z1mW299DdO6N4H4iQmMKOw3AY9juNiCoZyaEyiaDDXdOS7ADioDM9HIBGzjYMAhgVwAou48BpKmaSryk6DSBk8D7yY2DYKRNHR5IxhkCQ-AzKwuAXwN%2B-GMGABuMBOKzqDFFfRiUByOyyZf7GBIEEFyyBAEgIhHZL%2BACABeYA7qUCli-CE14ICUDzurABmtthTwqLPbA88OpL0MOLS6DMfAADoLAUL5mBLQmAiJODIBQbAsA6IQGDMmWAAB1AwsAiwAB1OBFm0FwnhlBYBrw3lvHeKRqBCKaAAfTAAAWngbwdRAAtPoTRhGZhUX0dQ4j0CSKsNIw%2BBCZ5z0sPfO6dklFFiaAABnUQATj6BwMxFjeH-WcQAIgYMI4R1DJBNAAHqBNCdrdQgTJBGBCWEiJkSjCxI0EYMx6AWFsJwJwuy3DLHEAADIqJEGjIgRB17wGccIkR6gDHIGEWIExORhHqHqWY3J7CCkgCKf4spFSqk1LqSIlxkThHYGESAEx4igA&content=IYYwLg9gTgBAwgVwM6QLYFMoCh0DsAmWWy6IwS6MAEsAQDboAKwAnhrmFgJAAOUAlqmBQWAQXDQYAXnjI0mbqDABlMOh4wkARhgAzKBFTdI3PITPEKZCjEYGQ6JEmXAG3fI5ACeYfhFzSMADkAHRhQVxc3HyCwmISsDKIKIYKXEqq6po6%2BobGEKYERFxQ6ABWEPy4mRraMADuABZ4hYRRXFUgdDXUtPgMzGx4nFwW6eA9SAAsUwBMAMw5BkZcJqNF6XQqahrAYNkw3FxNLZFj49tZ2sAA%2Bkt5kWvrbRZmQA)
